python/whylogs/core/relations.py

```diff
@@ -0,0 +1,167 @@
+        raise NotImplementedError
+        
+    @abstractmethod
+    def serialize(self) -> str:
+        raise NotImplementedError
+
+
+class LiteralGetter(ValueGetter):
+    def __init__(self, value: Union[str, int, float]) -> None:
+        self._literal = value
+
+    def __call__(self) -> Union[str, int, float]:
+        return self._literal
+
+    def serialize(self) -> str:
+        if isinstance(self._literal, str):
+            return f'"{self._literal}"'
+        return str(self._literal)
+
+
+class Relation(Enum):
+    match = 1
+    fullmatch = 2
+    equal = 3
+    less = 4
+    leq = 5
+    greater = 6
+    geq = 7
+    neq = 8
+    _and = 9
+    _or = 10
+    _not = 11
+    _udf = 12
+
+_TOKEN = ["", "~", "~=", "==", "<", "<=", ">", ">=", "!=", "and", "or", "not", "udf"]
+
+
+class Predicate:
+    def __init__(
+        self,
+        op: Relation,
+        value: Union[str, int, float, ValueGetter] = 0,
+        udf: Optional[Callable[[Any],bool]] = None,
+        left: Optional["Predicate"] = None,
+        right: Optional["Predicate"] = None,
+    ):
+        self._op = op
+        self._udf = udf
+        if op in {Relation.match, Relation.fullmatch}:
+            if not isinstance(value, str):
+                raise ValueError("match requires a string regular expression argument")
+            self._re = re.compile(value)
+
+        if isinstance(value, (str, int, float)):
+            self._value = LiteralGetter(value)
+        else:
+            self._value = value
+
+        if op in {Relation._and, Relation._or}:
+            if not left and right:
+                raise ValueError("boolean operators require two operands")
+        self._left = left
+        self._right = right
+
+        if op == Relation._not:
+            if not right:
+                raise ValueError("negation operator requires a predicate to negate")
+            self._right = right
+
+    def __call__(self, x: Any) -> bool:
+        op = self._op
+        if op == Relation.match:
+            return isinstance(x, str) and self._re.match(x)
+        if op == Relation.fullmatch:
+            return isinstance(x, str) and self._re.fullmatch(x)
+
+        value = self._value()
+        if op == Relation.equal:
+            return x == value
+        if op == Relation.less:
+            return x < value
+        if op == Relation.leq:
+            return x <= value
+        if op == Relation.greater:
+            return x > value
+        if op == Relation.geq:
+            return x >= value
+        if op == Relation.neq:
+            return x != value
+        if op == Relation._udf:
+            return self._udf(x)
+        if op == Relation._and:
+            return self._left(x) and self._right(x)
+        if op == Relation._or:
+            return self._left(x) or self._right(x)
+        if op == Relation._not:
+            return not self._right(x)
+
+        raise ValueError("Unknown ConditionCountMetric predicate")
+
```

Comment:
Are we happy with these names?

Reply:
I'm not a fan of leq cause I have to rethink it every time. less_eq is worth it to me. (And for all the others)

Reply:
I was referring to the fluent factory methods, not the `Relation` enumeration values since those aren't user-visible.

Reply:
Ooh ok, this didn't make sense where it was pointing at. GOt it. I do like those :) I'm happy with them